### ISSUE TITLE
How about robustness of image modification?

### ISSUE THREAD

**Issue by mychina75** [hurunyi/MaskMark#3]
Thank you for your great work.
It is works fine for message embedding. But after I crop part of the image, the decoded message broken.

![Image](https://github.com/user-attachments/assets/0d8ed8aa-72bb-4a16-914f-b1ae85287b87)
input: HELLO!!!
no change output: HELLO!!!
after change output: Aê+ãi'c

**Comment by hurunyi** [comment]
Thank you for your feedback and for testing the method!

Could you kindly share more details about how you performed the cropping (e.g., size and position of the cropped region)?

In our tests, the method is generally robust to partial crops, so we'd love to better understand the exact setup to help diagnose the issue.

**Comment by mychina75** [comment]
I tried to crop a square area from output image and resize to (256,256). The decoded message is not correct. 
To avoid crop partial of mask area, I further tested following pair. The crop area could cover the mask around area. 
result message is changed too. 
![Image](https://github.com/user-attachments/assets/86238f10-4284-4ec9-b8fc-daba911db6f3)
![Image](https://github.com/user-attachments/assets/80a72985-1813-46bf-b9a7-ae03f612edf0)

**Comment by mychina75** [comment]
Please check following process. maybe I made something wrong during the process.

1. message string encode to tensor array as following. then run your encoder codes.
#######################################
    string = "HELLO!!!"
    binary = [format(ord(char),'b') for char in string] #string_to_binary(string)
    binary_array = np.zeros(image.shape[0], message_length)
    for i in range(binary_array.size):
        id1 = i // 8
        id2 = i % 8
        binary_array[0][i] = binary[id1].zfill(8)[id2]
    message = torch.Tensor(binary_array).to(device)
########################################

2. Add a Ugly function to convert from tensor to string.
########################################
def binary_to_string(binary_list):
    str_array = [] 
    result = []
    for i in range(len(binary_list)):
        if i > 0 and i % 8 == 0:
            out = ''.join(str_array)
            number = int(out, 2)
            ch = chr(number)
            result.append(ch)  #print(ch)
            str_array = []
        if binary_list[i] < 0.5:
            str_array.append('0')
        else:
            str_array.append('1')
    return ''.join(result)
##########################################

3. crop area from watermarked image and save as 'wm_1703-1.png'. decode the message from it.
btw, no crop can get correct result.
###########################################
    image = Image.open("results/ED_64bits/wm_1703-1.png")
    image = transform(image).unsqueeze(0).to(device)
    wm_image_fused_256 = TF.resize(image, [256, 256])
    decoded_message, mask_pred_256 = wm_model.decoder(wm_image_fused_256)

    result = binary_to_string(decoded_message[0])
    print(result)
###########################################

**Comment by hurunyi** [comment]
Thanks for the clarification and the code you shared, the issue is now well understood and resolved on our side.

At first, we suspected there might be issues with your testing code. However, after reviewing the code you provided, we confirmed that everything looks correct on your end. Upon further inspection, we realized that we initially misunderstood the nature of the issue.

The failure in watermark extraction is actually caused by a compound distortion: Crop and Resize. This type of operation not only changes the spatial position of each pixel (geometric distortion) but also alters the pixel values due to interpolation during resizing (valuemetric distortion). It can thus be considered a new form of combined geometric + valuemetric distortion.

We tested the robustness of our provided models under this distortion and found that the embedded watermark in the original model indeed cannot withstand this kind of attack. The reason is that this specific distortion was not included in the distortion layer during training.

However, as described in our paper, the framework is designed to be flexible, and any custom distortion can be easily incorporated into the distortion layer for targeted robustness enhancement. Over the past few days, we fine-tuned the D_32bits model by including this Crop+Resize distortion in the training process. The fine-tuned model shows significant improvement in robustness under this compound distortion.

We have uploaded the fine-tuned model at https://huggingface.co/Runyi-Hu/MaskMark/resolve/main/D_32bits_crop%26resize_ft.pth?download=true,  feel free to download it and test on your side. If you encounter any further issues, we’d be happy to continue the discussion. Looking forward to your feedback!

**Comment by mychina75** [comment]
Tried you new model, works fine, great job~
would you please finetune all your models with Crop+Resize robustness? maybe more watermark attack methods.

**Comment by mychina75** [comment]
I made more tests, looks like the robustness for rotation is not good enough.

**Comment by mychina75** [comment]
Nice work~
I will try your new scripts.
